providers/gpgpu/bin/rvs.py

```diff
@@ -0,0 +1,223 @@
+#!/usr/bin/env python3
+"""ROCm Validation Suite wrapper.
+
+Copyright (C) 2024 Canonical Ltd.
+
+Authors
+  Pedro Avalos Jimenez <pedro.avalosjimenez@canonical.com>
+
+This program is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License version 3,
+as published by the Free Software Foundation.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+The purpose of this script is to simply wrap the ROCm Validation Suite
+executable, adding an appropriate failure exit code when a test fails.
+"""
+
+import argparse
+import logging
+import os
+import re
+import subprocess
+from pathlib import Path
+
+RVS_BIN = Path("/opt/rocm/bin/rvs")
+"""Default location for ROCm Validation Suite binary."""
+
+PLAINBOX_PROVIDER_DATA = Path(os.getenv("PLAINBOX_PROVIDER_DATA", "."))
+"""Location of the RVS module configurations."""
+
+
+def which_rvs() -> Path:
+    """Finds the location of the ROCm Validation Suite binary."""
+    proc = subprocess.run(
+        ["which", "rvs"],
+        check=False,
+        stdout=subprocess.PIPE,
+        universal_newlines=True,
+    )
+    return Path(proc.stdout.strip()) if proc.returncode == 0 else RVS_BIN
+
+
+class ModuleRunner:
```

Comment:
I think the whole ModuleRunner class could be simplified and made more explicit, getting rid of all the `subprocess.run` wrappers and the complexity in the validators:
```
class ModuleRunner:
    """This class represents the base module runner."""

    def __init__(self, rvs: Path, config: Path) -> None:
        """Initializes the module runner."""
        self.config = config
        self.rvs = rvs or Path(shutil.which("rvs") or RVS_BIN)

    def run(self, module: str):
        logging.debug("%s: RUNNING", module)

        result = subprocess.run(
            [self.rvs, "-c", self.config],
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            universal_newlines=True,
        )

        if result.returncode != 0:
            logging.error(result.stdout)
            logging.error(result.stderr)
            raise SystemExit("%s: FAILURE - failed to execute" % module)

        # Log the error output as debug
        if result.stderr:
            logging.debug(result.stderr)

        if hasattr(self, "_validate_output"):
            if not self._validate_output(result.stdout, module):
                logging.error(result.stdout)
                raise SystemExit("%s: FAILURE - validation failed" % module)
            
        logging.info(result.stdout)
        logging.debug("%s: SUCCESS", module)


class PassFailModuleRunner(ModuleRunner):
    """This class represents a module runner that passes or fails."""

    def _validate_output(self, output: str, module: str):
        # Find any of the common success messages in stdout
        for line in output.splitlines():
            if any(
                re.search(pass_str, line)
                for pass_str in [
                    r"%s true" % re.escape(module),
                    r"pass: TRUE",
                    r"GFLOPS \d+ Target GFLOPS: \d+ met: TRUE",
                ]
            ):
                return False
        return True


class MemModuleRunner(ModuleRunner):
    """This class represents the memory test module runner."""

    def _validate_output(self, output: str, module: str):
        return not all(
            "mem Test %s : PASS" % i in output.splitlines()
            for i in range(1, 12)
        )
```